### ISSUE TITLE
HttpClient connection should be resumed when response ends

### ISSUE THREAD

**Issue by steven-aerts** [eclipse/vert.x#2593]
We are running into a situation where an http client connection is paused when its last message is received.
This causes the paused connection to be recycled.
We think paused connections should be resumed when they are pooled.

A test to reproduce this behaviour:
```java
@RunWith(VertxUnitRunner.class)
public class PausedPooledConnectionTest {
    @Test
    public void test(TestContext context) {
        Vertx vertx = Vertx.vertx();
        vertx.createHttpServer()
                .requestHandler(req -> req.response().end("pong"))
                .listen(0, "localhost", context.asyncAssertSuccess(server -> {
                    HttpClient client = vertx.createHttpClient(new HttpClientOptions()
                            .setKeepAlive(true) // When set to false everything works as expected (first socket is also closed correctly)
                            .setDefaultPort(server.actualPort())
                    );
                    Async async = context.async();
                    // first request
                    client.get("/").handler(resp -> resp.endHandler(v -> {
                        resp.pause();
                        // second request which reuses the just paused connection.
                        vertx.setTimer(10, l -> client.get("/").handler(resp1 -> resp1.endHandler(v1 -> {
                            async.countDown();
                            vertx.close();
                        })).end());
                    })).end();
                }));
    }
}
```

A fix could be as easy as:
```diff
--- a/src/main/java/io/vertx/core/http/impl/Http1xClientConnection.java
+++ b/src/main/java/io/vertx/core/http/impl/Http1xClientConnection.java
@@ -874,6 +874,7 @@ class Http1xClientConnection extends Http1xConnectionBase implements HttpClientC
 
   private void recycle() {
     long expiration = keepAliveTimeout == 0 ? 0L : System.currentTimeMillis() + keepAliveTimeout * 1000;
+    doResume();
     listener.onRecycle(expiration);
   }
 }
```

If you think this is the correct fix, I do not mind to create a PR with the test and the fix.

**Comment by vietj** [comment]
this code has been partly rewritten, I'll keep the fix and port it myself to the branch.

**Comment by steven-aerts** [comment]
Wonderful let me know if I can do anything.

Thanks

**Comment by steven-aerts** [comment]
This one should work in `Http1xTest.java`.
While writing the test, I validated it on 3.5.3 where it fails as expected.
But on the master branch, this test is successful.  I tried to pin it down, and seems to be resolved between 03747cc and aa40d9b where you updated the http pooling behavior.  Nice.

```java
  @Test
  public void pausedPooledConnectionTest() throws Exception {
    server.requestHandler(req -> req.response().end("ok"));
    startServer();
    client.close();
    client = vertx.createHttpClient(new HttpClientOptions().setKeepAlive(true));
    client.getNow(DEFAULT_HTTP_PORT, DEFAULT_HTTP_HOST, DEFAULT_TEST_URI, resp1 -> {
      resp1.endHandler(v -> resp1.pause());
      vertx.setTimer(10, timerId -> client.getNow(DEFAULT_HTTP_PORT, DEFAULT_HTTP_HOST, DEFAULT_TEST_URI,
        resp2 -> resp2.endHandler(v -> testComplete())
      ));
    });
    await();
  }
```

Feel free to copy the code above or ask me to submit a PR.  I have signed the ECA in the past.

**Comment by vietj** [comment]
it seems to pass with latest vertx master

**Comment by vietj** [comment]
I'm curious : have you tried to call resume() on resp1 after a timer set in the same end handler to see if that works ? so something like

```
@Test
  public void pausedPooledConnectionTest() throws Exception {
    server.requestHandler(req -> req.response().end("ok"));
    startServer();
    client.close();
    client = vertx.createHttpClient(new HttpClientOptions().setKeepAlive(true));
    client.getNow(DEFAULT_HTTP_PORT, DEFAULT_HTTP_HOST, DEFAULT_TEST_URI, resp1 -> {
      resp1.endHandler(v -> {
         resp1.pause();
         vertx.setTimer(10, timeId -> resp1.resume());
      });
      vertx.setTimer(10, timerId -> client.getNow(DEFAULT_HTTP_PORT, DEFAULT_HTTP_HOST, DEFAULT_TEST_URI,
        resp2 -> resp2.endHandler(v -> testComplete())
      ));
    });
    await();
  }
```

**Comment by vietj** [comment]
that being said, if you call `pause()` you should call `resume()`, that's why I'm asking if calling resume() should work and indeed it works.

in master we have rewritten some part of the http client response handling so pause() is handled differently than in 3.5

**Comment by vietj** [comment]
ah, indeed now I think the main issue is that if that's paused the connection should not be recycled and the client should open a new connection

**Comment by vietj** [comment]
I think the main issue is that when the end handler is called, the pause/resume operations should be no-op

**Comment by vietj** [comment]
thanks for the relevant test

**Comment by vietj** [comment]
fix was good too, but we want to resume in response end and not in recycle

**Comment by steven-aerts** [comment]
Thanks for fixing this in 3.5.x too.

Wonderful.

**Comment by vietj** [comment]
yes I thought it is good to fix it, although we haven't planed a 3.5 release

**Comment by vietj** [comment]
as work around I advise to always resume the response when you pause it and never let a paused stream indefinitely
